dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/service/DefaultTrackerImportAccessManager.java

```diff
@@ -0,0 +1,254 @@
+        OrganisationUnitService organisationUnitService )
+    {
+        checkNotNull( aclService );
+        checkNotNull( ownershipAccessManager );
+        checkNotNull( organisationUnitService );
+
+        this.aclService = aclService;
+        this.ownershipAccessManager = ownershipAccessManager;
+        this.organisationUnitService = organisationUnitService;
+    }
+
+    @Override
+    public void checkReadEnrollmentAccess( ValidationErrorReporter reporter, User user,
+        ProgramInstance programInstance )
+    {
+        Objects.requireNonNull( user, USER_CANT_BE_NULL );
+        Objects.requireNonNull( programInstance, PROGRAM_INSTANCE_CANT_BE_NULL );
+        Objects.requireNonNull( programInstance.getProgram(), PROGRAM_CANT_BE_NULL );
+
+        checkProgramReadAccess( reporter, user, programInstance.getProgram() );
+
+        // TODO: can we use isReg instead?
+        if ( !programInstance.getProgram().isWithoutRegistration() )
+        {
+            checkTeiTypeAndTeiProgramAccess( reporter, user, programInstance.getEntityInstance(),
+                programInstance.getProgram() );
+        }
+        else
+        {
+            OrganisationUnit ou = programInstance.getOrganisationUnit();
+            if ( ou != null && !organisationUnitService.isInUserSearchHierarchyCached( user, ou ) )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1101 )
+                    .addArg( user )
+                    .addArg( ou ) );
+            }
+        }
+    }
+
+    protected void checkTeiTypeAndTeiProgramAccess( ValidationErrorReporter reporter, User user,
+        TrackedEntityInstance trackedEntityInstance,
+        Program program )
+    {
+        //TODO: TeiType
+        if ( !aclService.canDataRead( user, program.getTrackedEntityType() ) )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1092 )
+                .addArg( user )
+                .addArg( program ) );
+        }
+
+        if ( !ownershipAccessManager.hasAccess( user, trackedEntityInstance, program ) )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1102 )
+                .addArg( user )
+                .addArg( trackedEntityInstance )
+                .addArg( program ) );
+        }
+    }
+
+    @Override
+    public void checkWriteEnrollmentAccess( ValidationErrorReporter reporter, User user, Program program,
+        ProgramInstance programInstance )
+    {
+        Objects.requireNonNull( user, USER_CANT_BE_NULL );
+        Objects.requireNonNull( programInstance, PROGRAM_INSTANCE_CANT_BE_NULL );
+        Objects.requireNonNull( program, PROGRAM_CANT_BE_NULL );
+        //TODO: Investigate, must all program have a tei type?
+        // Objects.requireNonNull( program.getTrackedEntityType(), TRACKED_ENTITY_CANT_BE_NULL );
+
+//        OrganisationUnit ou = programInstance.getOrganisationUnit();
+//        // TODO: Investigate programInstance without ou possible or bug?
```

Comment:
if it exist, it's invalid data. not sure how we should deal with it, but we might consider it as if the data doesnt exist? @Bekkalizer @abyot